lib/http.go

```diff
@@ -80,10 +82,41 @@ func DecodeParams(r *http.Request, p interface{}) error {
 			return fmt.Errorf("unable to decode params from request body: %w", err)
 		}
 	}
+
+	if err := listParamsFromRequest(r, p); err != nil {
+		return fmt.Errorf("unable to pull list params from request url: %w", err)
+	}
+
 	// allow empty params
 	return nil
 }
 
+func listParamsFromRequest(r *http.Request, p interface{}) error {
+	pa, ok := p.(params.ListParams)
+	if !ok {
+		return nil
+	}
+	lp, err := pa.ListParamsFromRequest(r)
+	if err != nil {
+		return err
+	}
```

Comment:
I could make a strong argument against having an interface `params.ListParams` &  just use reflection to determine if a struct embeds the `params.List` struct. However, reflection is pretty costly, so I figured it might be better to use a less costly route (type assertion) to determine whether we even want to move on to using reflection in the first place.

But this is also a bit odd, so looking for feedback here.

Reply:
Yeah, I don't think the `ListParams` interface is really pulling its weight. For one, the method `ListParamsFromRequest` doesn't really use its owner at all, `p` in this case. Rather, it just wants to know if `p` has an embedded `params.List`, and if so, the relevant fields from the request should be assigned to it. I'm pretty sure you could do a type check on p instead of needing reflection. Something like:

```
  lp, ok := p.(params.List)
  if !ok {
    return
  }
  lp.limit = ...
```

Furthermore, reflection is only "slow" when compared to accessing fields directly. It's going to perform similarly to type assertions. Plus, if there's only a small number of reflection calls per http request, for example only like 4 or 5, the http overhead is going to dominate the runtime.

Reply:
Thank you this is so beyond valuable.

I set up a little go playground script to see how using type assertion to get at embedded fields would work, and it looks like it doesn't let us cast structs as the embedded field types: https://go.dev/play/p/_g50UfRvEMt

Your advice about the speed of reflection is very helpful. It makes me more confident that the right call is to nix the interface, and just use reflection to judge if something has list params.

Thank youuuuuuu